### ISSUE TITLE
Unexpected mutable struct copy when assigning a property of the struct

### ISSUE THREAD

**Issue by Meowtimer** [dotnet/roslyn#62563]
**Version Used**: 
main (9 Jul 2022)

**Steps to Reproduce**:
1. See https://sharplab.io/#gist:8b2bb153c0c0ab521af9c7e6febac02a

**Expected Behavior**:
The struct property always gets assigned in-place

**Actual Behavior**:
A copy of the struct is made only in the case where the right side is a switch expression.

**Comment by jaredpar** [comment]
The minimal repro here is essentially the following: 

```csharp
var val = 0.1;
var obj = new Obj();
ThrowWhenNull(obj).Color.Value = val switch
{
    _ => "red"
};
Console.WriteLine(obj.Color.Value);
ThrowWhenNull(obj).Color.Value = "green";
Console.WriteLine(obj.Color.Value);

static T ThrowWhenNull<T>(T obj) => obj ?? throw new System.InvalidOperationException();

struct Wrapper<T>
{
	private T _value;
	public T Value
	{
		get => _value;
		set => _value = value;
	}
}

class Obj
{
	public Wrapper<string> Color;
}
```

The issue is that `ThrowWhenNull(obj).Color` has different IL depending on the context here. In one case it uses a `ldfld` (copy) and the other we use `ldflda` (reference). This seems like a bug as I can't see a reason why we'd do this.

**Comment by RikkiGibson** [comment]
The left side of the assignment needs to be spilled before the switch expression as indicated here:

https://github.com/dotnet/roslyn/blob/5095dad0e6abccd0c6c659790b97d71b7331f2d7/src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_SwitchExpression.cs#L17-L25

Therefore we spill `ThrowWhenNull(obj).Color` to a temp. We have some logic for deciding whether the temp should be by-value or by-reference:

https://github.com/dotnet/roslyn/blob/5095dad0e6abccd0c6c659790b97d71b7331f2d7/src/Compilers/CSharp/Portable/Lowering/SpillSequenceSpiller.cs#L875-L884

In this case `CanBePassedByReference` is returning `false`, but it seems like it should return true--after all the expression `ref ThrowWhenNull(obj).Color` is valid. (Also, I don't think we should determine the writability of the temp field just based on whether the expression type is readonly struct. It seems like a readonly field should also require us to take a readonly reference. However, since all the value-checks have occurred and passed by this point, I don't see why it should even matter here. Maybe it affects whether we make copies in some case?)

I think this bit in `CanBePassedByReference` is wrong:

https://github.com/dotnet/roslyn/blob/a528fe2d1721466c4ada3e97bd99a70521560f58/src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs#L922-L928

Here we basically assume a field can only be passed by ref if its receiver can be passed by ref. Since `ref ThrowWhenNull(obj)` is disallowed, we return `false` from `CanBePassedByReference`. But `ref ThrowWhenNull(obj).Color`, being a field of a reference type is totally valid. We do have to be careful that `ref ThrowWhenNull(someValueType).SomeField` is *not* valid though.

I would have thought we would determine this answer similarly to [`HasHome`](https://github.com/dotnet/roslyn/blob/91aa5211754abfd5441072e8c2ffc0f9a321e300/src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs#L4041) (full impl too large to embed), where if the field is not readonly and not on a value type then we just say, yes it has a writable home. 

Basically, I think the fix here is to just say that if a field is on a reference type, you can take a reference to it.
